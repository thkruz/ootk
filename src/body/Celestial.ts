/**
 * @author @thkruz Theodore Kruczek
 * @license AGPL-3.0-or-later
 * @copyright (c) 2020-2024 Theodore Kruczek
 *
 * Orbital Object ToolKit is free software: you can redistribute it and/or modify it under the
 * terms of the GNU Affero General Public License as published by the Free Software
 * Foundation, either version 3 of the License, or (at your option) any later version.
 *
 * Orbital Object ToolKit is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY;
 * without even the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
 * See the GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General Public License along with
 * Orbital Object ToolKit. If not, see <http://www.gnu.org/licenses/>.
 */

import { AzEl, Degrees, Kilometers, RAD2DEG, RaDec, Radians, Sun } from '../main.js';

/**
 * Celestial is a static class that provides methods for calculating the position of celestial objects such as the Sun,
 * Moon, and planets in the sky.
 */
export class Celestial {
  private constructor() {
    // disable constructor
  }

  /**
   * Calculates the azimuth and elevation of a celestial object at a given date, latitude,
   * longitude, right ascension, and declination.
   * @param date - The date for which to calculate the azimuth and elevation.
   * @param lat - The latitude of the observer.
   * @param lon - The longitude of the observer.
   * @param ra - The right ascension of the celestial object.
   * @param dec - The declination of the celestial object.
   * @returns An object containing the azimuth and elevation in degrees.
   */
  static azEl(date: Date, lat: Degrees, lon: Degrees, ra: Radians, dec: Radians): AzEl<Degrees> {
    const c: RaDec = {
      ra,
      dec,
      dist: 0 as Kilometers,
    };
    const azEl = Sun.azEl(date, lat, lon, c);

    const el = <Radians>(azEl.el + Celestial.atmosphericRefraction(azEl.el)); // elevation correction for refraction

    return {
      az: (azEl.az * RAD2DEG) as Degrees,
      el: (el * RAD2DEG) as Degrees,
    };
  }

  /**
   * Atmospheric refraction in astronomy, refers to the bending of light as it passes through the Earth's
   * atmosphere. This effect is most noticeable for celestial objects like stars and planets when they are
   * close to the horizon. Here's a breakdown of how it works:
   *
   * Actual Position: Due to this bending of light, the apparent position of a celestial object is slightly
   * different from its true position in the sky. When a star or planet is near the horizon, the effect is more
   * pronounced because the light path passes through more of the Earth's atmosphere, which increases the amount of
   * bending.
   *
   * A familiar example of atmospheric refraction is observed during sunrise and sunset. The Sun appears to
   * be above the horizon when it is actually just below it. This is because the light from the Sun is bent
   * upwards as it passes through the atmosphere.
   * @param h - elevation
   * @returns refraction
   */
  static atmosphericRefraction(h: Radians): Radians {
    if (h < 0) {
      h = <Radians>0;
    }

    return <Radians>(0.0002967 / Math.tan(h + 0.00312536 / (h + 0.08901179)));
  }

  /**
   * Calculate the declination. Similar to latitude on Earth, declination is another celestial coordinate.
   * It measures how far north or south an object is from the celestial equator
   * @param l - ecliptic longitude
   * @param b - ecliptic latitude
   * @returns declination
   */
  static declination(l: number, b: number): Radians {
    return <Radians>Math.asin(Math.sin(b) * Math.cos(Sun.e) + Math.cos(b) * Math.sin(Sun.e) * Math.sin(l));
  }

  /**
   * Calculate the right ascension. This is a celestial coordinate used to determine the position of objects
   * in the sky. It's analogous to longitude on Earth. Right Ascension indicates how far east an object is
   * from the vernal equinox along the celestial equator.
   * @param l - ecliptic longitude
   * @param b - ecliptic latitude
   * @returns right ascension
   */
  static rightAscension(l: number, b: number): Radians {
    return <Radians>Math.atan2(Math.sin(l) * Math.cos(Sun.e) - Math.tan(b) * Math.sin(Sun.e), Math.cos(l));
  }

  /**
   * Calculate the elevation. Elevation, or altitude, is the angle between an object in the sky and the
   * observer's local horizon. It's commonly expressed in degrees, where 0 degrees is right at the horizon
   * and 90 degrees is directly overhead (the zenith), but we are using radians to support trigonometric
   * functions like Math.sin() and Math.cos().
   * @param H - siderealTime
   * @param phi - latitude
   * @param dec - The declination of the sun
   * @returns elevation
   */
  static elevation(H: number, phi: Radians, dec: Radians): Radians {
    return <Radians>Math.asin(Math.sin(phi) * Math.sin(dec) + Math.cos(phi) * Math.cos(dec) * Math.cos(H));
  }

  /**
   * Calculate the azimuth. This is a compass direction measurement. Azimuth measures the angle along
   * the horizon from a specific reference direction (usually true north) to the point where a vertical
   * line from the object intersects the horizon.
   * @param H - siderealTime
   * @param phi - latitude
   * @param dec - The declination of the sun
   * @returns azimuth in rad
   */
  static azimuth(H: number, phi: Radians, dec: Radians): Radians {
    return <Radians>(Math.PI + Math.atan2(Math.sin(H), Math.cos(H) * Math.sin(phi) - Math.tan(dec) * Math.cos(phi)));
  }
}
