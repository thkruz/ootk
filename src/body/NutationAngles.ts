/**
 * @author @thkruz Theodore Kruczek
 * @license AGPL-3.0-or-later
 * @copyright (c) 2020-2024 Theodore Kruczek
 *
 * Orbital Object ToolKit is free software: you can redistribute it and/or modify it under the
 * terms of the GNU Affero General Public License as published by the Free Software
 * Foundation, either version 3 of the License, or (at your option) any later version.
 *
 * Orbital Object ToolKit is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY;
 * without even the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
 * See the GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General Public License along with
 * Orbital Object ToolKit. If not, see <http://www.gnu.org/licenses/>.
 */

import { Radians } from '../main.js';

/** Represents the nutation angles. */
export type NutationAngles = {
  /** The nutation in longitude (Δψ) in radians. */
  dPsi: Radians;
  /** The nutation in obliquity (Δε) in radians. */
  dEps: Radians;
  /** The mean obliquity of the ecliptic (ε₀) in radians. */
  mEps: Radians;
  /** The true obliquity of the ecliptic (ε) in radians. */
  eps: Radians;
  /** The equation of the equinoxes (ΔΔt) in radians. */
  eqEq: Radians;
  /** The Greenwich Apparent Sidereal Time (GAST) in radians. */
  gast: Radians;
};
