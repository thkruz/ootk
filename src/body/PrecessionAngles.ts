import { Radians } from '../ootk';

// / Earth precession angles _(rad)_.
export type PrecessionAngles = {
  zeta: Radians;
  theta: Radians;
  zed: Radians;
};
