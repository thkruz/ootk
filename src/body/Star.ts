import { AzEl, Degrees, RaDec, Radians } from '@src/ootk';
import { Sun } from './Sun';

export class Star {
  private constructor() {
    // disable constructor
  }

  static getStarAzEl(date: Date, lat: Degrees, lon: Degrees, ra: Radians, dec: Radians): AzEl<Radians> {
    const c: RaDec = {
      ra,
      dec,
    };
    const azEl = Sun.azEl(date, lat, lon, c);

    const el = <Radians>(azEl.el + Star.astroRefraction(azEl.el)); // elevation correction for refraction

    return {
      az: azEl.az,
      el,
    };
  }

  /**
   * get astro refraction
   * @param {Radians} h - elevation
   * @returns {number} refraction
   */
  static astroRefraction(h: Radians): Radians {
    if (h < 0) {
      h = <Radians>0;
    }

    return <Radians>(0.0002967 / Math.tan(h + 0.00312536 / (h + 0.08901179)));
  }
}
