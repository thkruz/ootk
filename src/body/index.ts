export { Celestial } from './Celestial';
export { Earth } from './Earth';
export { Moon } from './Moon';
export { Sun } from './Sun';
