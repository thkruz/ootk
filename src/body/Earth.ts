/**
 * @author @thkruz Theodore Kruczek
 * @license AGPL-3.0-or-later
 * @copyright (c) 2020-2024 Theodore Kruczek
 *
 * Orbital Object ToolKit is free software: you can redistribute it and/or modify it under the
 * terms of the GNU Affero General Public License as published by the Free Software
 * Foundation, either version 3 of the License, or (at your option) any later version.
 *
 * Orbital Object ToolKit is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY;
 * without even the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
 * See the GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General Public License along with
 * Orbital Object ToolKit. If not, see <http://www.gnu.org/licenses/>.
 */

import {
  angularDiameter,
  AngularDiameterMethod,
  asec2rad,
  DataHandler,
  DEG2RAD,
  earthGravityParam,
  EpochUTC,
  evalPoly,
  Kilometers,
  RAD2DEG,
  Radians,
  RadiansPerSecond,
  secondsPerDay,
  secondsPerSiderealDay,
  TAU,
  ttasec2rad,
  Vector3D,
} from '../main.js';
import { NutationAngles } from './NutationAngles.js';
import { PrecessionAngles } from './PrecessionAngles.js';

// / Earth metrics and operations.
export class Earth {
  private constructor() {
    // disable constructor
  }

  // / Earth gravitational parameter _(km²/s³)_.
  static readonly mu: number = earthGravityParam;

  // / Earth equatorial radius.
  static readonly radiusEquator = 6378.1363 as Kilometers;

  // / Earth coefficient of flattening _(unitless)_.
  static readonly flattening: number = 1.0 / 298.257223563;

  // / Earth polar radius.
  static readonly radiusPolar = (Earth.radiusEquator * (1.0 - Earth.flattening)) as Kilometers;

  // / Earth mean radius.
  static readonly radiusMean = ((2.0 * Earth.radiusEquator + Earth.radiusPolar) / 3.0) as Kilometers;

  // / Earth eccentricity squared _(unitless)_.
  static readonly eccentricitySquared: number = Earth.flattening * (2.0 - Earth.flattening);

  // / Earth J2 effect coefficient _(unitless)_.
  static readonly j2: number = 1.08262668355315e-3;

  // / Earth J3 effect coefficient _(unitless)_.
  static readonly j3: number = -2.53265648533224e-6;

  // / Earth J4 effect coefficient _(unitless)_.
  static readonly j4: number = -1.619621591367e-6;

  // / Earth J5 effect coefficient _(unitless)_.
  static readonly j5: number = -2.27296082868698e-7;

  // / Earth J6 effect coefficient _(unitless)_.
  static readonly j6: number = 5.40681239107085e-7;

  // / Earth rotation vector _(rad/s)_.
  static readonly rotation = new Vector3D<RadiansPerSecond>(
    0 as RadiansPerSecond,
    0 as RadiansPerSecond,
    7.292115146706979e-5 as RadiansPerSecond,
  );

  // / Calculate mean motion _(rad/s)_ from a given [semimajorAxis] _(km)_.
  static smaToMeanMotion(semimajorAxis: Kilometers): RadiansPerSecond {
    return Math.sqrt(Earth.mu / (semimajorAxis * semimajorAxis * semimajorAxis)) as RadiansPerSecond;
  }

  /**
   * Converts revolutions per day to semi-major axis.
   * @param rpd - The number of revolutions per day.
   * @returns The semi-major axis value.
   */
  static revsPerDayToSma(rpd: number): number {
    return Earth.mu ** (1 / 3) / ((TAU * rpd) / secondsPerDay) ** (2 / 3);
  }

  // / Calculate Earth [PrecessionAngles] at a given UTC [epoch].
  static precession(epoch: EpochUTC): PrecessionAngles {
    const t = epoch.toTT().toJulianCenturies();
    const zeta = evalPoly(t, Earth.zetaPoly_);
    const theta = evalPoly(t, Earth.thetaPoly_);
    const zed = evalPoly(t, Earth.zedPoly_);

    return { zeta: zeta as Radians, theta: theta as Radians, zed: zed as Radians };
  }

  // / Calculate Earth [NutationAngles] for a given UTC [epoch].
  static nutation(epoch: EpochUTC): NutationAngles {
    const t = epoch.toTT().toJulianCenturies();
    const moonAnom = evalPoly(t, Earth.moonAnomPoly_);
    const sunAnom = evalPoly(t, Earth.sunAnomPoly_);
    const moonLat = evalPoly(t, Earth.moonLatPoly_);
    const sunElong = evalPoly(t, Earth.sunElongPoly_);
    const moonRaan = evalPoly(t, Earth.moonRaanPoly_);
    let deltaPsi = 0.0;
    let deltaEpsilon = 0.0;
    const dh = DataHandler.getInstance();

    for (let i = 0; i < 4; i++) {
      const [a1, a2, a3, a4, a5, ai, bi, ci, di] = dh.getIau1980Coeffs(i);
      const arg = a1 * moonAnom + a2 * sunAnom + a3 * moonLat + a4 * sunElong + a5 * moonRaan;
      const sinC = ai + bi * t;
      const cosC = ci + di * t;

      deltaPsi += sinC * Math.sin(arg);
      deltaEpsilon += cosC * Math.cos(arg);
    }
    deltaPsi *= ttasec2rad;
    deltaEpsilon *= ttasec2rad;
    const meanEpsilon = evalPoly(t, Earth.meanEpsilonPoly_);
    const epsilon = meanEpsilon + deltaEpsilon;
    const eqEq =
      deltaPsi * Math.cos(meanEpsilon) +
      0.00264 * asec2rad * Math.sin(moonRaan) +
      0.000063 * asec2rad * Math.sin(2.0 * moonRaan);
    const gast = epoch.gmstAngle() + eqEq;

    return {
      dPsi: deltaPsi as Radians,
      dEps: deltaEpsilon as Radians,
      mEps: meanEpsilon as Radians,
      eps: epsilon as Radians,
      eqEq: eqEq as Radians,
      gast: gast as Radians,
    };
  }

  // / Convert a [semimajorAxis] _(km)_ to an eastward drift rate _(rad/day)_.
  static smaToDrift(semimajorAxis: number): number {
    const t = (TAU * Math.sqrt(semimajorAxis ** 3 / Earth.mu)) / secondsPerSiderealDay;

    return (1.0 - t) * TAU;
  }

  // / Convert a [semimajorAxis] _(km)_ to an eastward drift rate _(°/day)_.
  static smaToDriftDegrees(semimajorAxis: number): number {
    return Earth.smaToDrift(semimajorAxis) * RAD2DEG;
  }

  // / Convert an eastward [driftRate] _(rad/day)_ to a semimajor-axis _(km)_.
  static driftToSemimajorAxis(driftRate: number): number {
    const t = (-driftRate / TAU + 1) * secondsPerSiderealDay;

    return ((Earth.mu * t * t) / (4 * Math.PI * Math.PI)) ** (1 / 3);
  }

  // / Convert an eastward [driftRate] _(°/day)_ to a semimajor-axis _(km)_.
  static driftDegreesToSma(driftRate: number): number {
    return Earth.driftToSemimajorAxis(DEG2RAD * driftRate);
  }

  /**
   * Calculates the diameter of the Earth based on the satellite position.
   * @param satPos The position of the satellite.
   * @returns The diameter of the Earth.
   */
  static diameter(satPos: Vector3D): number {
    return angularDiameter(Earth.radiusEquator * 2, satPos.magnitude(), AngularDiameterMethod.Sphere);
  }

  // / Earth precession `zeta` polynomial coefficients.
  private static zetaPoly_: Float64Array = Float64Array.from([
    0.017998 * asec2rad,
    0.30188 * asec2rad,
    2306.2181 * asec2rad,
    0.0,
  ]);

  // / Earth precession `theta` polynomial coefficients.
  private static thetaPoly_: Float64Array = Float64Array.from([
    -0.041833 * asec2rad,
    -0.42665 * asec2rad,
    2004.3109 * asec2rad,
    0.0,
  ]);

  // / Earth precession `zed` polynomial coefficients.
  private static zedPoly_: Float64Array = Float64Array.from([
    0.018203 * asec2rad,
    1.09468 * asec2rad,
    2306.2181 * asec2rad,
    0,
  ]);

  private static moonAnomPoly_: Float64Array = Float64Array.from([
    1.4343e-5 * DEG2RAD,
    0.0088553 * DEG2RAD,
    (1325.0 * 360.0 + 198.8675605) * DEG2RAD,
    134.96340251 * DEG2RAD,
  ]);

  // / Earth nutation Sun anomaly polynomial coefficients.
  private static sunAnomPoly_: Float64Array = Float64Array.from([
    3.8e-8 * DEG2RAD,
    -0.0001537 * DEG2RAD,
    (99.0 * 360.0 + 359.0502911) * DEG2RAD,
    357.52910918 * DEG2RAD,
  ]);

  // / Earth nutation Moon latitude polynomial coefficients.
  private static moonLatPoly_: Float64Array = Float64Array.from([
    -2.88e-7 * DEG2RAD,
    -0.003542 * DEG2RAD,
    (1342.0 * 360.0 + 82.0174577) * DEG2RAD,
    93.27209062 * DEG2RAD,
  ]);

  // / Earth nutation Sun elongation polynomial coefficients.
  private static sunElongPoly_: Float64Array = Float64Array.from([
    1.831e-6 * DEG2RAD,
    -0.0017696 * DEG2RAD,
    (1236.0 * 360.0 + 307.1114469) * DEG2RAD,
    297.85019547 * DEG2RAD,
  ]);

  // / Earth nutation Moon right-ascension polynomial coefficients.
  private static moonRaanPoly_: Float64Array = Float64Array.from([
    2.139e-6 * DEG2RAD,
    0.0020756 * DEG2RAD,
    -(5.0 * 360.0 + 134.1361851) * DEG2RAD,
    125.04455501 * DEG2RAD,
  ]);

  // / Earth nutation mean epsilon polynomial coefficients.
  private static meanEpsilonPoly_: Float64Array = Float64Array.from([
    0.001813 * asec2rad,
    -0.00059 * asec2rad,
    -46.815 * asec2rad,
    84381.448 * asec2rad,
  ]);
}
