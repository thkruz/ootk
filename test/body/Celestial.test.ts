import { Degrees, Radians } from './../../src/types/types';
import { Celestial } from './../../src/body/Celestial';

describe('Celestial', () => {
  /*
   * Calculates the azimuth and elevation of a celestial object at a given date,
   * latitude, longitude, right ascension, and declination.
   */
  it('should calculate the azimuth and elevation when given valid inputs', () => {
    const mockExampleDate = new Date(1705109326817);
    const lat = 37.7749 as Degrees; // latitude of San Francisco
    const lon = -122.4194 as Degrees; // longitude of San Francisco
    const ra = 1.3963 as Radians; // right ascension of a celestial object
    const dec = 0.7854 as Radians; // declination of a celestial object

    const azEl = Celestial.azEl(mockExampleDate, lat, lon, ra, dec);

    expect(azEl).toMatchSnapshot();
  });

  // Calculates the atmospheric refraction for a given elevation.
  it('should calculate the atmospheric refraction when given a valid elevation', () => {
    const elevation = 0.5 as Radians;
    const refraction = Celestial.atmosphericRefraction(elevation);

    expect(refraction).toMatchSnapshot();
  });

  // Calculates the declination of a celestial object given its ecliptic longitude and latitude.
  it('should calculate the declination when given valid ecliptic longitude and latitude', () => {
    const longitude = 0.5 as Radians;
    const latitude = 0.3 as Radians;

    const declination = Celestial.declination(longitude, latitude);

    expect(declination).toMatchSnapshot();
  });

  // Calculates the right ascension of a celestial object given its ecliptic longitude and latitude.
  it('should calculate the right ascension when given valid ecliptic longitude and latitude', () => {
    const longitude = 0.5 as Radians;
    const latitude = 0.3 as Radians;

    const rightAscension = Celestial.rightAscension(longitude, latitude);

    expect(rightAscension).toMatchSnapshot();
  });

  // Calculates the elevation of a celestial object given its sidereal time, latitude, and declination.
  it('should calculate the elevation when given valid sidereal time, latitude, and declination', () => {
    const siderealTime = 1.2;
    const latitude = 0.5 as Radians;
    const declination = 0.3 as Radians;

    const elevation = Celestial.elevation(siderealTime, latitude, declination);

    expect(elevation).toMatchSnapshot();
  });

  // Calculates the azimuth of a celestial object given its sidereal time, latitude, and declination.
  it('should calculate the azimuth when given valid sidereal time, latitude, and declination', () => {
    const siderealTime = 1.2;
    const latitude = 0.5 as Radians;
    const declination = 0.3 as Radians;

    const azimuth = Celestial.azimuth(siderealTime, latitude, declination);

    expect(azimuth).toMatchSnapshot();
  });

  // Calculates the atmospheric refraction for an elevation of 0.
  it('should calculate the atmospheric refraction when elevation is 0', () => {
    const elevation = 0 as Radians;

    const refraction = Celestial.atmosphericRefraction(elevation);

    expect(refraction).toMatchSnapshot();
  });
});
