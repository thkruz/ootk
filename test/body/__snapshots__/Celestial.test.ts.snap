// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Celestial should calculate the atmospheric refraction when elevation is 0 1`] = `0.008446689093277518`;

exports[`Celestial should calculate the atmospheric refraction when given a valid elevation 1`] = `0.0005363221501954175`;

exports[`Celestial should calculate the azimuth and elevation when given valid inputs 1`] = `
Object {
  "az": 58.636843073419904,
  "el": 39.53055812197483,
}
`;

exports[`Celestial should calculate the azimuth when given valid sidereal time, latitude, and declination 1`] = `4.816878749628938`;

exports[`Celestial should calculate the declination when given valid ecliptic longitude and latitude 1`] = `0.4704906087409695`;

exports[`Celestial should calculate the elevation when given valid sidereal time, latitude, and declination 1`] = `0.46170567615036917`;

exports[`Celestial should calculate the right ascension when given valid ecliptic longitude and latitude 1`] = `0.3464473691313705`;
